src/main/java/hw25/jdbc/LessonDao.java

```diff
@@ -0,0 +1,132 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+public class LessonDao {
+
+    public Lesson insert(final Lesson lesson) {
+        Objects.requireNonNull(lesson);
+        if (lesson.getId() != null) {
+            throw new JdbcOperationException("ID mustn't be provided during th insert operation!");
+        }
+        String sql = """
+                WITH new_homework AS (
+                    INSERT INTO homework (name, description) VALUES (?, ?) RETURNING id
+                    )
+                INSERT INTO lesson (name, homework_id) SELECT ?, id FROM new_homework
+                RETURNING lesson.id, lesson.updated_at, (SELECT id FROM new_homework)
+                """;
+        try (PreparedStatement prepStatement = DataBaseConnection.getConnection()
+                .prepareStatement(sql)) {
+            int setIndex = 1;
+            prepStatement.setString(setIndex++, lesson.getHomework().getName());
+            prepStatement.setString(setIndex++, lesson.getHomework().getDescription());
+            prepStatement.setString(setIndex, lesson.getName());
+            ResultSet resultSet = prepStatement.executeQuery();
+            if (resultSet.next()) {
+                int getIndex = 1;
+                lesson.setId(resultSet.getLong(getIndex++));
+                lesson.setTimeStamp(resultSet.getTimestamp(getIndex++).toLocalDateTime());
+                lesson.getHomework().setId(resultSet.getLong(getIndex));
+            } else {
+                throw new JdbcOperationException("No rows were inserted!");
+            }
+            return lesson;
+        } catch (SQLException e) {
+            throw new JdbcOperationException("Failed to insert the lesson!", e);
+        }
+    }
+
+    public String delete(final Long id) {
+        Objects.requireNonNull(id);
+        String sql = """
+                WITH deleted_homework AS (DELETE FROM lesson WHERE id = ?
+                  RETURNING homework_id)
+                DELETE FROM homework WHERE id IN (SELECT homework_id FROM deleted_homework)
+                 AND NOT EXISTS (SELECT 1 FROM lesson WHERE homework_id = 
+                  (SELECT homework_id FROM deleted_homework) AND id <> ?)
+                """;
+        try (PreparedStatement prepStatement = DataBaseConnection.getConnection()
+                .prepareStatement(sql)) {
+            int index = 1;
+            prepStatement.setLong(index++, id);
+            prepStatement.setLong(index, id);
+            int result = prepStatement.executeUpdate();
+            if (result == 1) {
+                return "Lesson and related homework were successfully deleted";
+            } else if (result == 0) {
+                return "Lesson was successfully deleted. Homework wasn't deleted because " +
+                        "other lesson has this homework also.";
+            } else {
+                throw new JdbcOperationException("No rows were deleted!");
+            }
```

Comment:
I describe this peace of code as 0 rows will be returned if no rows were deleted. So this if-else can be simplified by using only `if (deletedRows < 1) { throw JdbcOperationException;}`. The rest can be omitted.